rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (lib.concatStringsSep "\n" (lib.mapAttrsToList (key: value: ''
+  // ${value.reason}
+  pref("${key}", ${builtins.toJSON value.value});
+'') defaultPrefs));
+```
+
+It is arguably pretty hard to read and reason about. Even when applying some more whitespace-generous formatting:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (
+  lib.concatStringsSep "\n" (
+    lib.mapAttrsToList
+    (
+      key: value: ''
+        // ${value.reason}
+        pref("${key}", ${builtins.toJSON value.value});
+      ''
+    )
+    defaultPrefs
+  )
+);
+```
+
+One can observe the following issues:
+
+- If you want to follow the data flow, you must read it from bottom to top,
+  from the inside to the outside (the input here is `defaultPrefs`).
+- Adding a function call to the output would require wrapping the entire
+  expression in parentheses and increasing its indentation.
+
+Compare this to the equivalent call with `lib.pipe`:
+
+```nix
+defaultPrefsFile = pipe defaultPrefs [
+  (lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  ))
+  (lib.concatStringsSep "\n")
+  (pkgs.writeText "nixos-default-prefs.js")
+];
+```
+
+The code now clearly reads from top to bottom in the order the data is processed,
+it is easy to add and remove processing steps at any point.
+
+With a dedicated pipe operator, it would look like this:
+
+```nix
+defaultPrefsFile = defaultPrefs
+  |> lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  )
+  |> lib.concatStringsSep "\n"
+  |> pkgs.writeText "nixos-default-prefs.js";
+```
+
+The artificial distinction between the first input and the functions via the list now is gone,
+and so are the parentheses around the functions.
+With the lower syntax overhead, using the operator becomes attractive in more situations,
+whereas a `pipe` pays for its overhead only in more complex scenarios (usually three functions or more).
+Having a dedicated operator also increases visibility and discoverability of the feature.
+
+# Detailed design
+[design]: #detailed-design
+
+## `|>` operator
+
+A new operator `|>` is introduced into the Nix language.
+Semantically, it is defined as the reverse of function application: `f a` = `a |> f`.
+It is left-associative and has a binding strength one weaker than function application:
+`a |> f |> g b |> h` = `h ((g b) (f a))`.
+
+## `builtins.pipe`
+
+`lib.pipe`'s functionality is implemented as a built-in function.
+The main motivation for this is that it allows to give better error messages
+like line numbers when some part of the pipeline fails.
+Additionally, it allows easy usage outside of Nixpkgs and increases discoverability.
+
+While Nixpkgs is bounds to minimum Nix versions and thus `|>` won't be available until
```

Comment:
I think this is a great argument against going through with this RFC.

Reply:
This would be an argument for never making any changes to improve the language at all, ever. Unless you see some shiny alternative language to which we'll realistically all migrate to within the next five years, any improvements to the Nix language are still expected to be beneficial.

I've actually thought about proposing this change for quite some time now, and the main thing that held me back was the idea is that Nix is a "lost cause" and that I should rather spending my energy towards some successor. But eventually, I realized that realistically, we'll be stuck with Nix for a long time to come, so better invest the energy to make it a bit more comfortable …

Also note that outside of Nixpkgs, the feature will be available almost immediately. There is a lot of Nix code outside of Nixpkgs, all these flakes, shells, and users' system configurations. Projects like home-manager, devenv, nur, etc. are all free to migrate to newer Nix versions at their own pace. We shouldn't ignore these.